features/329-principal-apis/design.md

```diff
@@ -0,0 +1,136 @@
+# Background
+Auth gear supports various identity providers, such as password credentials and
+OAuth. Auth gear should support all identity providers equally well, and
+provide a consistent API for manipulating principals.
+
+
+# Use Cases
+- Get which method is used to log in the current session
+- List all methods to authenticated as a user
+- Add/remove user authentication method
+- Send welcome email to OAuth accounts
+- Populate user metadata according to external profile
+- Create password credentials for users signed up with OAuth
+
+
+# Glossary
+
+**Identity Provider**:
+A method to assert the identity of user. For example, providing a password or
+perform OAuth flow.
+
+**Principal**:
+An identity associated with user. A user can have multiple principals. Since
+the term 'Principal' is defined differently across many frameworks and
+standards, we will instead use a simpler term 'Identity' for public interfaces.
+
+
+# Proposed Design
+
+## Re-authentication for Security
+Many principal manipulation operations are critical for security:
+- Add/remove login ID
+- Link/unlink SSO provider
+- Change password
+
+By default, user are required to re-authenticate before performing these
+operations. This is achieve by checking the issue time of the access token is
+recent.
+
+For change password, we support the common case of providing old password to
+implicitly authenticate the user.
+
+Both the re-auth requirement and time to considered 'recent' can be configured.
+Developer can turn off the re-auth requirement in configuration. However,
+developer should understand the potential security risk.
+
+> Related config: `reauthForSecurity` and `reauthIntervalSeconds`
+> Related API: `changePassword` function
```

Comment:
add/remove login id, link/unlink sso provider?

Reply:
API for SSO is specified in another document, so it is not mentioned here.

Reply:
However it is mentioned in the paragraph above; I assume if "changePassword" is related APIs, do you meant any APIs that need to be reauthenticated? That seems include link/unlink SSO and add/remove Login ID? In that case why `changePassword` is listed as related APIs but not the others?

I have written another comment on review about the difficult to read reasons and recommendation.